reps/2023-10-30-gpu-memory-support.md

```diff
@@ -0,0 +1,208 @@
+
+- https://discuss.ray.io/t/how-to-specify-gpu-resources-in-terms-of-gpu-ram-and-not-fraction-of-gpu/4128
+
+- https://discuss.ray.io/t/gpu-memory-aware-scheduling/2922/5
+
+- https://discuss.ray.io/t/automatic-calculation-of-a-value-for-the-num-gpu-param/7844/4
+
+- https://discuss.ray.io/t/ray-train-ray-tune-ray-clusters-handling-different-gpus-with-different-gpu-memory-sizes-in-a-ray-cluster/9220
+
+This REP allows users to directly schedule fractional GPU resources by amount of GPU memory. In our design, if a user specifies `gpu_memory = 20GB`, then `ray` automatically converts the value to `num_gpus` depending on which node the request is assigned to. As example, if it's scheduled on A100 40GB node, then `num_gpus = 0.5`, otherwise if it's scheduled on A100 80GB node, then `num_gpus = 0.25`. As a result, user can schedule a fixed amount of GPU resources without depending on which types of GPUs the tasks/actos are scheduled to.
+
+### Should this change be within `ray` or outside?
+
+Inside `ray` project since we want to add new parameter `gpu_memory` to the `ray` remote function.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@pcmoritz, @jjyao, @scv119
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@jjyao
+
+## Design
+
+### API
+Users will be able to specify amount of GPU memory to their ray tasks/actors using `gpu_memory` on `ray.remote`. The specified `gpu_memory` will be the amount of GPU resources from a single GPU that will be allocated to users ray tasks/actors.
+
+```python
+# Request a fractional GPU with specified gpu_memory in bytes.
+# Mutually exclusive with num_gpus.
+@ray.remote(gpu_memory=1024 * 1024 * 1024) # 1 mb request
+def task:
+  …
+```
+
+When a Ray node is started, Ray will auto detect the number of GPUs and GPU memory of each GPU. Users also have the option to manually specify them:
+
+```bash
+ray start # auto detection
+
+ray start –num_gpus=3 –gpu_memory=1000 * 1024 * 1024 * 1024 # manual override, each GPU has 1000mb memory
+```
+
+Note that GPU memory and GPU is 1-1 conversion, means 20GB of GPU memory is equivalent to 0.5 fractional value of an `A100_40GB` GPU. So, for simplicity and consistency, ray doesn't allow users to specify both `num_gpus` and `gpu_memory` in a single ray task/actor.
+
+
+```python
+# Request a fractional GPU both num_gpus and gpu_memory is not allowed
+@ray.remote(gpu_memory=1024 * 1024 * 1024, num_gpus=0.5) # raise ValueError exception
```

Comment:
Could they both be allowed? If both `num_gpus` and `gpu_memory` are specified, then it would require that much memory on that many GPUs. `num_gpus` would default to 1, so not specifying it would get the behavior described above. It could be an error condition to specify a fractional value for `num_gpus` if also specifying `gpu_memory`. Thoughts?